packages/hash/search-loader/src/searchLoader.ts

```diff
@@ -0,0 +1,213 @@
+      await this.processNextQueueMsg(1000);
+    }
+    this.isStopped = true;
+  }
+
+  /** Process the next item on the queue, or return early if the queue is empty for
+   * longer than `timeout` milliseconds. */
+  private async processNextQueueMsg(timeout: number): Promise<void> {
+    const processed = await this.queueConsumer.pop(
+      this.searchQueueName,
+      timeout,
+      async (item: string) => {
+        const wal2jsonMsg = JSON.parse(item) as Wal2JsonMsg;
+        await this.loadMsgIntoSearchIndex(wal2jsonMsg);
+        logger.debug(item);
+        return true;
+      },
+    );
+    if (processed) {
+      this.statsd?.increment("messages_processed");
+    }
+  }
+
+  /** Load a message into the search index. */
+  private async loadMsgIntoSearchIndex(wal2jsonMsg: Wal2JsonMsg) {
+    const table = wal2jsonMsg.table;
+    if (table === "entity_versions") {
+      const entity = EntityVersion.parseWal2JsonMsg(wal2jsonMsg);
+
+      const entityType = await EntityType.getEntityType(this.db, {
+        entityTypeVersionId: entity.entityTypeVersionId,
+      });
+
+      if (!entityType) {
+        logger.error(
+          `Could not find entity type with entityTypeVersionId "${entity.entityTypeVersionId}"`,
+        );
+        return;
+      }
+
+      const indexedEntity: EntitiesDocument = {
+        accountId: entity.accountId,
+        entityId: entity.entityId,
+        entityVersionId: entity.entityVersionId,
+        entityTypeId: entityType.entityId,
+        entityTypeVersionId: entity.entityTypeVersionId,
+        entityTypeName: entityType.metadata.name,
+        updatedAt: entity.updatedAt.toISOString(),
+        updatedByAccountId: entity.updatedByAccountId,
+      };
+      // @todo: could move the `SYSTEM_TYPES` definition from the backend to backend-utils
+      // and use it here rather than checking the string value of the entity type name.
+      if (
+        entityType.accountId === this.systemAccountId &&
+        entityType.metadata.name === "Text"
+      ) {
+        indexedEntity[ENTITIES_SEARCH_FIELD] = textEntityPropertiesToFTS(
+          entity.properties,
+        );
+        // @todo this should be done through a model class instead of explicitly through the DB adapter.
+        const grandparents = await this.db.getAncestorReferences({
+          accountId: entity.accountId,
+          entityId: entity.entityId,
+          depth: 2,
+        });
+        // @todo: Do we handle text blocks that have multiple grandparent pages?
+        if (grandparents.length === 1) {
+          const grandparentLatestEntity = await Entity.getEntityLatestVersion(
+            this.db,
+            {
+              ...grandparents[0],
+            },
+          );
+          const pageSystemTypeName: SystemType = "Page";
+
+          if (
+            grandparentLatestEntity &&
+            grandparentLatestEntity.entityType.metadata.name ===
+              pageSystemTypeName
+          ) {
```

Comment:
**Blocking: ** this is not a robust way of determining whether a page is a page entity on the BE, as users could fool this by creating an entity type with the name `"Page"` in their namespace.

We should compare the IDs of the entity types instead (using `(await Page.getEntityType(...)).entityId` for instance)

Reply:
We could load the page system type once when we `start` the search loader. What do you think about that instead of fetching it every time?

Reply:
We can do that yes, you're right this will speed things up now although I'm not sure it's really worth the effort as long-term we'll definitely want some caching solution for _all_ system types. In general I've not worried about calls to `Page.getEntityType` for that reason.

Reply:
Yes good point. That has been mentioned before. I don't think it's an issue to do this basic caching here right now. 

I can make the call directly to the model class system type method if you prefer, though.

Reply:
I'm happy for us to merge the temporary caching solution you've implemented, just want to make sure that we're on the same page that this should be removed once we have a generalised caching solution in place.

Reply:
Yes, and I appreciate that moving to model class methods will help us enable generalized caching too. It makes a lot of sense :)